src/Perl6/Metamodel/ContainerSpecProtocol.nqp

```diff
@@ -0,0 +1,34 @@
+role Perl6::Metamodel::ContainerSpecProtocol {
+    has $!code_pair;
+
+    method get_container_spec($obj) {
+        $!code_pair
+    }
+
+    method set_container_spec($obj, $code_pair) {
+        $!code_pair := $code_pair;
+    }
+
+    method publish_container_spec($obj) {
+        # If we have a container specification here, install that
+        if $!code_pair {
+            nqp::setcontspec($obj, 'code_pair', $!code_pair);
+            $obj.HOW.compose_repr($obj);
+        }
+
+        # look in the parents for any if none here
+        else {
+            for self.mro($obj) -> $class {
+                if nqp::can($class.HOW, 'get_container_spec') {
+                    my $code_pair := $class.HOW.get_container_spec($class);
+                    if $code_pair {
+                        $!code_pair := $code_pair;
```

Comment:
I don't think it should be assigned here, otherwise we lose track of which class it was actually set on (not terribly important, but still, feels odd to do this).

Reply:
Addressed with 963646644f

BTW, for some reason, I couldn't get:

    if $class.HOW.get_container_spec($class) -> $code_pair {

to work: it would fail in compilation with:

    Too few positionals passed; expected 3 arguments but got 2
       at gen/moar/Metamodel.nqp:298  (blib/Perl6/Metamodel.moarvm:publish_container_spec)